### PR TITLE
update brute_force delete algorithm to avoid re-using ids

### PR DESCRIPTION
**Pull request by meiravgri**
## Changes in DS:
1. Removed **VectorBlockMember** DS, it is no longer needed.

2.  **VectorBlock** 
- removed **member** mapping array, as the block and relative index of a vector can be simply calculated according to:
 block = id / block_size
 relative_index = id % block_size

3. **BruteForceIndex**
- **deletedIds** is also redundant
- **idToVectorBlockMemberMapping** changed to idToLabelMapping so that idToLabelMapping[id] = LabelToId[id]

## Algorithm changes:
- previous algorithm: storing deleted ids and trying to re-use them upon new vector insertion
- current algorithm: move the label & data of the last vector in the index to the deleted vector place.

## Bug fix:
Wrong bytes count to copy when updating a vector data.

### CODE REVIEW

**Review by GuyAv46 — COMMENTED** [review]
Nice work! its about time we get rid of this `blockMembers`

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 1]
```diff
@@ -43,24 +42,27 @@ size_t BruteForceIndex::estimateInitialSize(const BFParams *params) {
...
+    return params->dim * sizeof(float) + sizeof(idType);
 }
 
 void BruteForceIndex::updateVector(idType id, const void *vector_data) {
-    // Get the vector block
-    VectorBlockMember *vectorBlockMember = this->idToVectorBlockMemberMapping[id];
-    VectorBlock *vectorBlock = vectorBlockMember->block;
-    size_t index = vectorBlockMember->index;
+
+    // TODO Get the vector block
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:60]: to remove the todo...

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 2]
```diff
@@ -73,103 +75,115 @@ int BruteForceIndex::addVector(const void *vector_data, size_t label) {
...
-                VectorBlock(this->vectorBlockSize, this->dim, this->allocator);
-            this->vectorBlocks.push_back(vectorBlock);
-        }
+        this->vectorBlocks.push_back(new_vectorBlock);
+    }
+
+    // get the last vectors block
+    VectorBlock *vectorBlock = this->vectorBlocks.back();
+
+    assert(vectorBlock == vectorBlocks[count / vectorBlockSize]);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:101]: yes

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 3]
```diff
@@ -73,103 +75,115 @@ int BruteForceIndex::addVector(const void *vector_data, size_t label) {
...
+
+    // add vector data to vectorBlock
+    vectorBlock->addVector(vector_data);
+
+    // if idToLabelMapping is full,
+    // resize and align idToLabelMapping by vectorBlockSize
+    size_t idToLabelMapping_size = this->idToLabelMapping.size();
+
+    if (count >= idToLabelMapping_size) {
+        size_t last_block_vectors_count = count % vectorBlockSize;
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:111]: K thought about it and i agree

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 4]
```diff
@@ -73,103 +75,115 @@ int BruteForceIndex::addVector(const void *vector_data, size_t label) {
...
+    assert(last_vector_block == vectorBlocks.at(last_idx / vectorBlockSize));
+
+    float *last_vector_data = last_vector_block->removeAndFetchVector();
+
+    // Get the vectorBlock and the relative index of the deleted id.
+    VectorBlock *deleted_vectorBlock = getVectorVectorBlock(id_to_delete);
+    size_t id_to_delete_rel_idx = getVectorRelativeIndex(id_to_delete);
+
+    // Put data of last vector inpalce of the deleted vector.
+    deleted_vectorBlock->updateVector(id_to_delete_rel_idx, last_vector_data);
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:167]: changed the order and added an if

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 5]
```diff
@@ -54,6 +60,27 @@ class BruteForceIndex : public VecSimIndex {
...
+size_t BruteForceIndex::getVectorRelativeIndex(idType id) {
+    // get the relative index of this id inside the vectorBlock
+    return id % vectorBlockSize;
+}
+
+labelType BruteForceIndex::getVectorLabel(idType id) const { return idToLabelMapping.at(id); }
+
+void BruteForceIndex::setVectorLabel(idType id, labelType new_label) {
+    idToLabelMapping.at(id) = new_label;
+}
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.h:86]: ?

File: tests/unit/test_allocator.cpp [hunk 6]
```diff
@@ -162,12 +156,14 @@ TEST_F(AllocatorTest, test_bf_index_block_size_1) {
     int deleteCommandAllocationDelta = VecSimIndex_DeleteVector(bfIndex, 2);
     expectedAllocationDelta -=
         (sizeof(VectorBlock) + vecsimAllocationOverhead); // Free the vector block
-    expectedAllocationDelta -= (sizeof(VectorBlockMember) + vecsimAllocationOverhead);
     expectedAllocationDelta -=
-        (sizeof(VectorBlockMember *) +
-         vecsimAllocationOverhead); // Pointer for the new vector block member
+        sizeof(VectorBlock *) + vecsimAllocationOverhead; // remove from vectorBlocks vector
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_allocator.cpp:160]: no :)

File: tests/unit/test_allocator.cpp [hunk 7]
```diff
@@ -162,12 +156,14 @@ TEST_F(AllocatorTest, test_bf_index_block_size_1) {
...
+    expectedAllocationDelta -=
+        sizeof(labelType) + vecsimAllocationOverhead; // resize idToLabelMapping
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_allocator.cpp:162]: no, and also no decrease in size...

**Review by GuyAv46 — DISMISSED** [review]
Nice work!

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 8]
```diff
@@ -43,24 +42,27 @@ size_t BruteForceIndex::estimateInitialSize(const BFParams *params) {
...
     // Get the vector block
-    VectorBlockMember *vectorBlockMember = this->idToVectorBlockMemberMapping[id];
-    VectorBlock *vectorBlock = vectorBlockMember->block;
-    size_t index = vectorBlockMember->index;
+    VectorBlock *vectorBlock = getVectorVectorBlock(id);
+    size_t index = getVectorRelativeIndex(id);
+
     // Update vector data in the block.
-    float *destinaion = vectorBlock->getVector(index);
-    memcpy(destinaion, vector_data, this->dim);
```
Comment (alonre24) [inline src/VecSim/algorithms/brute_force/brute_force.cpp:63]: This was a bug! Good catch! Also, I would point to this fix in the PR description...

File: tests/unit/test_bruteforce.cpp [hunk 9]
```diff
@@ -31,11 +31,58 @@ TEST_F(BruteForceTest, brute_force_vector_add_test) {
...
     }
     VecSimIndex_AddVector(index, (const void *)a, 1);
+
     ASSERT_EQ(VecSimIndex_IndexSize(index), 1);
+
     VecSimIndex_Free(index);
 }
 
-TEST_F(BruteForceTest, resizeIndex) {
+TEST_F(BruteForceTest, brute_force_vector_update_test) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_bruteforce.cpp:40]: please also check that after the update the blob in the index was changed correctly

File: src/VecSim/algorithms/brute_force/brute_force.cpp (hunk 8 above)
Comment (GuyAv46) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:63]: and add in the tests that the new vector blob was updated correctly

Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:63]: tested in:
- test_delete_swap_block 
- brute_force_vector_update_test

File: tests/unit/test_bruteforce.cpp [hunk 10]
```diff
@@ -1315,7 +1545,7 @@ TEST_F(BruteForceTest, testCosine) {
 
 TEST_F(BruteForceTest, testSizeEstimation) {
     size_t dim = 128;
-    size_t n = 10000;
+    size_t n = 0;
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce.cpp:1548]: why this change?

File: tests/unit/test_bruteforce.cpp [hunk 11]
```diff
@@ -31,11 +31,104 @@ TEST_F(BruteForceTest, brute_force_vector_add_test) {
...
+    // Add another vector and try to remove the first.
+    size_t last_vec_label = 2;
+    VecSimIndex_AddVector(index, (const void *)a, last_vec_label);
+    ASSERT_EQ(VecSimIndex_IndexSize(index), 2);
+    // Since we reached the limit of the initial capacity(1) id2label is
+    // increased and aligned with vector block size.
+    ASSERT_EQ(reinterpret_cast<BruteForceIndex *>(index)->idToLabelMapping.size(), bs);
+
+    size_t first_vec_label = 1;
+    size_t deleted_id = 0;
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce.cpp:93]: isn't it this same logic as the swap test?
also, please rename `deleted_id` to `first_vec_new_id` or something similar. This is a bit confusing

**Review by DvirDukhan — COMMENTED** [review]
very good

File: src/VecSim/algorithms/brute_force/brute_force.cpp [hunk 12]
```diff
@@ -43,24 +42,27 @@ size_t BruteForceIndex::estimateInitialSize(const BFParams *params) {
     est += (params->metric == VecSimMetric_L2 ? sizeof(L2Space) : sizeof(InnerProductSpace)) +
            sizeof(size_t);
     // Parameters related part.
-    est += params->initialCapacity * sizeof(decltype(idToVectorBlockMemberMapping)::value_type) +
-           sizeof(size_t);
+
+    if (params->initialCapacity) {
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply src/VecSim/algorithms/brute_force/brute_force.cpp:46]: when initial size is 0 there is no allocation overhead (the size_t part)

File: tests/unit/test_bruteforce.cpp (hunk 11 above)
Comment (meiravgri) [reply tests/unit/test_bruteforce.cpp:93]: you are right. Swap test purpose is the to ensure the indexes of vector inside a block dont change upon deletion and this test also checks removing the last vector special case. ill move deletion of last vector to the delete swap test.

File: tests/unit/test_bruteforce.cpp (hunk 10 above)
Comment (meiravgri) [reply tests/unit/test_bruteforce.cpp:1548]: To check if the estimation wont include the allocation overhead of id2labelMapping.
size estimation with an initial capacity is covered in the next test

File: tests/unit/test_bruteforce.cpp [hunk 13]
```diff
@@ -31,11 +31,104 @@ TEST_F(BruteForceTest, brute_force_vector_add_test) {
...
+
+    // VectorBlocks vector is empty.
+    ASSERT_EQ(reinterpret_cast<BruteForceIndex *>(index)->vectorBlocks.size(), 0);
+    // id2label size is also empty.
+    ASSERT_EQ(reinterpret_cast<BruteForceIndex *>(index)->idToLabelMapping.size(), 0);
+
+    // Label2id of the last vector doesn't exist.
+    ASSERT_EQ(reinterpret_cast<BruteForceIndex *>(index)->labelToIdLookup.find(last_vec_label),
+              reinterpret_cast<BruteForceIndex *>(index)->labelToIdLookup.end());
+
```
[the remark this answers is not in the document]
Comment (meiravgri) [reply tests/unit/test_bruteforce.cpp]: checked right after calling delete

File: .circleci/config.yml [hunk 14]
```diff
@@ -188,7 +188,7 @@ jobs:
 
   build-macos:
       macos:
-        xcode: 11.6.0
+        xcode: 11.7.0
```
Comment (DvirDukhan) [inline .circleci/config.yml:191]: remove as this is not part of this PR

File: tests/unit/test_bruteforce.cpp [hunk 15]
```diff
@@ -533,7 +504,8 @@ TEST_F(BruteForceTest, brute_force_reindexing_same_vector_different_id) {
 }
 
 TEST_F(BruteForceTest, test_delete_swap_block) {
-    size_t n = 6;
+    size_t n = 6;                 // idToLabelMapping initial size.
+    size_t initiail_capacity = 5; // idToLabelMapping initial size.
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce.cpp:508]: 5 or 6?

Comment (meiravgri) [reply tests/unit/test_bruteforce.cpp]: moved n above the loop and removed the comment

**Review by DvirDukhan — APPROVED** [review]
